### ISSUE TITLE
Anchor generation

### ISSUE THREAD

**Issue by krishnak** [zhen8838/K210_Yolo_framework#7]
I am trying to understand the Anchor generation algorithm.

I have created a custom model and I am using yolo v3 style detection. 

I looked at some of the keras yolo v3 implementations, they are all having anchors as integers.

some of those anchors are bigger than the input image size as I am using 224x224 input

https://github.com/qqwweee/keras-yolo3/blob/master/model_data/yolo_anchors.txt

However I read some where on AlexyAB github repository that anchors have to be generated taking in to consideration of the model input size. 

Can you please explain the logic of your code in generating anchors as you are producing a floating point anchor. 

My training annotation file is in this format as below, single file, single line per image. If I want to modify your code for Kmeans generation, what all should I be modifying to get the correct anchors.

```
racoon_dataset/images/images/raccoon-115.jpg 51,130,351,556,1
racoon_dataset/images/images/raccoon-116.jpg 51,130,351,556,1
```

My other question is yolo v3 has 3 anchors per detection level, in your voc_anchor.np you have only 6 anchor pairs, are you using only 2 detection levels?

**Comment by zhen8838** [comment]
Yes, K210 mobilenet yolov3 using only 2 detection layers. 

Because the value of anchor will become floating-point in the process of calculating loss, I directly fix the range of anchor value between 0 and 1.

You can refer to the following two blogs:

https://zhen8838.github.io/2019/07/10/yolo-error
https://zhen8838.github.io/2019/03/12/yolo-anchor

**Comment by krishnak** [comment]
Thank you for replying, I was using google translate to read it, but its difficult to understand as translation isn't great.

I will run through individual code line by line and try to understand.

**Comment by krishnak** [comment]
I have another question, as per yolo, the output of a detector is NxNx (5+1)*number of classes.

If I have a trained network to detect 10 classes, my model outputs will be NxNx60

If I want to detect an extra class(10+1) my detector output will be NxNx66

I have already trained for 10 class detection, now if I want to detect this 1 extra class, how do I do it? because this change affects all the detectors, but detectors can't detect if backbone is not trained. Do I just train using the new class images with the existing model by replacing the  detectors alone to N/[32:16:8]xN/[32:16:8]x66 ? if I replace the detectors won't it loose the previously learnt classes.

**Comment by zhen8838** [comment]
I use keras model to training, so you can use :
```python
old_weights = old_model.weights
new_weights = new_model.weights

for old_weight, new_weight in zip(old_weights, new_weights):
  # you can choice weights which you want to load
  new_weight.assign(old_weight)
```

**Comment by krishnak** [comment]
Thank you, I haven't used your technique of loading mobilenet weights yet, I trained from scratch to see how it performs, as I was only using the racoon/kangaroo image sets which were small. After about 1000 iterations the loss is below 10 but doesn't decrease much. I stopped training and tried to infer with the weights.

As you can see from the inference image, the bounding box is drawn beyond the image co-ordinates. This is mainly due to the the y+h/2  having a value > 1, my code at decoding input is,

can you let me know whether my bounding box code logic is correct and if it is correct whether the bounding box errors will disappear with further training?
```
x, y, w, h = netout[int(row)][int(col)][b][:4]
            x = (col + x) / grid_w # center position, unit: image width
            y = (row + y) / grid_h # center position, unit: image height
            w = anchors[2 * b + 0] * np.exp(w) / net_w # unit: image width
            h = anchors[2 * b + 1] * np.exp(h) / net_h # unit: image height  
            if y+h/2 > 1:
                print('anchor is ',anchors[2 * b + 1]) 
            # last elements are class probabilities
            classes = netout[int(row)][col][b][5:]
            box = BoundBox(x-w/2, y-h/2, x+w/2, y+h/2, objectness, classes)
```
![boxscaling](https://user-images.githubusercontent.com/1259447/76674005-e5217700-65d0-11ea-882d-cda0e7d431aa.png)
